gcloud-java-datastore/src/test/java/com/google/gcloud/datastore/DatastoreTest.java

```diff
@@ -445,7 +483,80 @@ public void testRunStructuredQuery() {
     assertEquals(20, entity.getLong("age"));
     assertEquals(1, entity.properties().size());
     assertFalse(results4.hasNext());
-    // TODO(ozarov): construct a test to verify nextQuery/pagination
+  }
+
+  @Test
+  public void testStructuredQueryPagination() {
+    DatastoreRpcFactory rpcFactoryMock = EasyMock.createStrictMock(DatastoreRpcFactory.class);
+    DatastoreRpc rpcMock = EasyMock.createStrictMock(DatastoreRpc.class);
+    EasyMock.expect(rpcFactoryMock.create(EasyMock.anyObject(DatastoreOptions.class)))
+        .andReturn(rpcMock);
+    List<com.google.datastore.v1beta3.RunQueryResponse> responses =
+        buildResponsesForQueryPagination();
+    for (int i = 0; i < responses.size(); i++) {
+      try {
+        EasyMock
+            .expect(rpcMock.runQuery(
+                EasyMock.anyObject(com.google.datastore.v1beta3.RunQueryRequest.class)))
+            .andReturn(responses.get(i));
+      } catch (DatastoreRpcException e) {
+        fail("Unexpected DatastoreRpcException");
+      }
+    }
+    EasyMock.replay(rpcFactoryMock, rpcMock);
+    DatastoreOptions options =
+        this.options.toBuilder()
+            .retryParams(RetryParams.getDefaultInstance())
+            .serviceRpcFactory(rpcFactoryMock)
+            .build();
+    Datastore mockDatastore = DatastoreFactory.instance().get(options);
+    QueryResults<Key> results = mockDatastore.run(Query.keyQueryBuilder().build());
+    int count = 0;
+    while (results.hasNext()) {
+      count += 1;
+      results.next();
+    }
+    assertEquals(count, responses.size());
+    EasyMock.verify(rpcFactoryMock, rpcMock);
+  }
+
+  private List<com.google.datastore.v1beta3.RunQueryResponse> buildResponsesForQueryPagination() {
```

Comment:
Nice, could we actually go through 3 pages (I want to make sure that the structured query that was created for gql pages correctly).